pkg/runtime/stack_trace.go

```diff
@@ -0,0 +1,165 @@
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package runtime
+
+import (
+	"fmt"
+	"runtime"
+	"strings"
+)
+
+type FramesIterator interface {
+	// Next works the same as standard (*runtime.Frames).Next().
+	Next() (runtime.Frame, bool)
+}
+
+type StackTrace interface {
+	// ProgramCounters copies a slice of program counters to the argument.
+	// The program counters which begins with the final caller and ends with the initial/root caller.
+	ProgramCounters(PCs) int
+
+	// Frames is an analog of runtime.CallerFrames.
+	Frames() FramesIterator
+
+	// String implements fmt.Stringer.
+	String() string
+
+	// Len returns the amount of frames in the stack trace
+	Len() int
+}
+
+// PCs is a goroutine execution frames stack trace.
+type PCs []PC
+
+var _ StackTrace = (PCs)(nil)
+
+// Len implements interface StackTrace.
+func (s PCs) Len() int {
+	return len(s)
+}
+
+// ProgramCounters implements interface StackTrace.
+func (s PCs) ProgramCounters(out PCs) int {
+	return copy(out, s)
+}
+
+// CallersFrames is an equivalent of standard runtime.CallerFrames.
+func (s PCs) Frames() FramesIterator {
+	pcs := make([]uintptr, len(s))
+	for idx, pc := range s {
+		pcs[idx] = uintptr(pc)
+	}
+	return runtime.CallersFrames(pcs)
+}
+
+// String implements fmt.Stringer.
+func (s PCs) String() string {
+	var result strings.Builder
+	frames := s.Frames()
+	if frames == nil {
+		return "<invalid stack trace>"
+	}
+	frameDepth := 1
+	for {
+		frame, ok := frames.Next()
+		result.WriteString(fmt.Sprintf("%d. %s:%d: %s\n", frameDepth, frame.File, frame.Line, frame.Function))
```

Comment:
According to the iterator, it is possible that an empty value is returned by function, which might cause a panic in this line. Since the object is an interface, is it possible to add an extra check to evaluate if the object is not nil.

Reply:
Sorry, could you elaborate?

`frame` is not an interface, but a structure: [`runtime.Frame`](https://pkg.go.dev/runtime#Frame); while `frames` was checked against `nil` above and it is already not nil by this moment.

Reply:
Should you not check `ok` before trying to access members of `frame`?

Reply:
No, this is bad naming. `ok` here means if `Next` should be called once more to get a new entry after that or this is the last entry.

Reply:
Updated the name, to make it more self-descriptive.